frame/bridge/crab/issuing/src/lib.rs

```diff
@@ -0,0 +1,118 @@
+//! # Crab Issuing Module
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+mod types {
+	// --- darwinia ---
+	use crate::*;
+
+	pub type MappedRing = u128;
+
+	pub type AccountId<T> = <T as frame_system::Trait>::AccountId;
+
+	pub type RingBalance<T> = <RingCurrency<T> as Currency<AccountId<T>>>::Balance;
+
+	type RingCurrency<T> = <T as Trait>::RingCurrency;
+}
+
+// --- substrate ---
+use frame_support::{
+	decl_error, decl_event, decl_module, decl_storage, ensure,
+	traits::{Currency, ExistenceRequirement, Get},
+};
+use frame_system::ensure_signed;
+use sp_runtime::{traits::AccountIdConversion, ModuleId, SaturatedConversion};
+// --- darwinia ---
+use types::*;
+
+pub trait Trait: frame_system::Trait {
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+
+	type ModuleId: Get<ModuleId>;
+
+	type RingCurrency: Currency<AccountId<Self>>;
+
+	/// Weight information for extrinsics in this pallet.
+	type WeightInfo: WeightInfo;
+}
+
+pub trait WeightInfo {}
+impl WeightInfo for () {}
+
+decl_event! {
+	pub enum Event<T>
+	where
+		AccountId = AccountId<T>,
+		RingBalance = RingBalance<T>,
+	{
+		/// Someone swapped some *CRING*. [who, swapped *CRING*, burned Mapped *RING*]
+		SwapAndBurn(AccountId, RingBalance, MappedRing),
+	}
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+			/// Swap Amount - TOO LOW
+			SwapAmountTL,
+			/// Backed *RING* - INSUFFICIENT
+			BackedRingIS,
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as DarwiniaCrabIssuing {
+		pub BackedRing
```

Comment:
BackedRing might have concept confusion
```suggestion
		pub TotalMappedRing
```